hashintel/hash
# Entity type editor – persist changes, and related updates

## 🌟 What is the purpose of this PR?




Persist updates to list of properties in the entity type editor, and related changes. See below

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203007126736604/1203144688196124/f) _(internal)_

## 🚫 Blocked by




N/A

## 🔍 What does this change?




Functionality changes:

- Always display the latest entity type version, not just version 1
- Make publish changes and discard changes buttons on entity type edit bar work
- New loading state for the edit bar while publishing, which disabled buttons. @nonparibus I'll need to consult on what these buttons should look like while disabled, not sure we have that design anywhere
- Display already persisted property types for an entity type when loading the editor

Code changes:

- Integrate the type system loading process into `useEntityType` so that `onComplete` isn't called until the type system has loaded (as we need the type system to be loaded for that). This is enabled by new hook `useAdvancedInitTypeSystem` so as to not mess with other consumers of `useInitTypeSystem`. This provides a function which you call to get a promise which will resolve (or already be resolved) when the type system has loaded.
- New `mustBeVersionedUri` to make it easier to handle strings we know to be versioned URIs already
- Integrate `react-hook-form` for storing the list of properties for an entity type
- New function to update the entity type, and replace the stored entity type version once that's completed
- Edit bar set up to not update values while animating its exit, using a new `useFrozenValue` hook
- Property types context is now a map, to make finding the right one easier

## 📜 Does this require a change to the docs?



- The docs for x need updating to explain that y

## ⚠️ Known issues



## 🐾 Next steps



## 🛡 What tests cover this?



-

## ❓ How to test this?



1.  Open the entity type editor
2. Trying making a change and clicking publish changes
3. Make sure the contents of the edit bar don't change while it animates away
4. Try making another change
5. Click discard, see reverts back to where you were after save
6. Refresh, see the changes persisted
7. Try making another change
8. Click discard, see reverts back to where you were after save

## 📹 Demo
